### ISSUE TITLE
Request to backport some new features/fixes of net.request and session to stable branches

### ISSUE THREAD

**Issue by zeeker999** [electron/electron#26549]
Our products reply on following commits from master branch and we have to backport these commits to previous stable branch(10-x-y) for our products internally.

0e7d59d #25284 feat: [net] add "credentials" option to net.request (#25284)
0fc5f18b #26134 fix: Don't sort the headers of ClientRequest (#26134)
201fc11b #24937 feat: Added support for all proxy modes (#24937)
e1cc78f2 #26135 feat: allow setting the Origin header and Sec-Fetch-* headers in net.request() (#26135)

We would like to upstream these changes and not want our products to depend on a custom build of electron. These new features are fully covered by unit tests and should be pretty safe. Is ok to backport these commits to both 10-x-y and 11-x-y branches, or the branch 11-x-y at least?

@nornagon @zcbenz

**Comment by ckerr** [comment]
@LuoJinghua, semver minor backports are usually voted on by the Releases working group. Happily enough, its next scheduled meeting is tonight. I've added the `backport/requested` label to this ticket to ensure that it's seen by the group.

**Comment by ckerr** [comment]
Reporting back from the @electron/wg-releases meeting: There was concern that #26134 and #26135 might be breaking changes due to the changing of the header ordering and the injection of new headers -- and so the question has been referred to @electron/wg-api to answer this question. :roller_coaster:

**Comment by LuoJinghua** [comment]
#26134 is a bugfix and make the behavior of net.request match other counterparts(e.g. http.request). I don't think it will break  anything. 

#26135 doesn't inject any new headers really. It make net.request respect the headers set by user. Without the commit, net.request hardcode following headers regardless whatever set by user and removes the header `Sec-Fetch-User` unconditionly.
```
'Sec-Fetch-Dest': 'empty',
'Sec-Fetch-Mode': 'no-cors',
'Sec-Fetch-Site': 'none'
```
And the behavior of net.request won't change if user doesn't set any of `Origin`, `Sec-Fetch-Dest`, `Sec-Fetch-Mode`, `Sec-Fetch-User` and `Sec-Fetch-Dest`.

The only possible change which need change from user is:
* `Sec-Fetch-Mode` is `cors`
* `CORS` rule is enabled for the request
* The `credentials` is `include`
* Credentials isn't allowed by the target server.

The credentials should set to `omit` for this case like we did in a browser env.

**Comment by LuoJinghua** [comment]
Sorry to ping you since this issue hasn't updated for a long time.
@electron/wg-api

**Comment by MarshallOfSound** [comment]
Sorry folks, I've been sitting on this one a while.  My perspective here (and this probably should have been caught before this landed) is that we can't backport this PR in particular as it is a breaking change.

> e1cc78f #26135 feat: allow setting the Origin header and Sec-Fetch-* headers in net.request() (#26135)

This PR adjusts the behavior of **an existing** option (the `headers` object) to send unexpected additional headers if you provide an `origin` header.  For `fetch` and the usecase of mirroring the `fetch` API behavior this makes sense, but for folks who are expecting the `net` API to more closely mirror the `http` and `https` APIs from Node.JS these additional headers being added and being unavoidable for existing code / requests is IMO a breaking change.

We could make this not-breaking by not inferring the `origin` option via the `origin` header but even then I'd want to test all existing requests to ensure we haven't accidentally started adding random headers to existing functioning requests.

TLDR: My stance is that adding unexpected and unremovable headers to a request is a semver/major change

**Comment by LuoJinghua** [comment]
> Sorry folks, I've been sitting on this one a while. My perspective here (and this probably should have been caught before this landed) is that we can't backport this PR in particular as it is a breaking change.
> 
> > [e1cc78f](https://github.com/electron/electron/commit/e1cc78f275054034a063422bbb1f53daf1c552ed) #26135 feat: allow setting the Origin header and Sec-Fetch-* headers in net.request() (#26135)
> 
> This PR adjusts the behavior of **an existing** option (the `headers` object) to send unexpected additional headers if you provide an `origin` header. For `fetch` and the usecase of mirroring the `fetch` API behavior this makes sense, but for folks who are expecting the `net` API to more closely mirror the `http` and `https` APIs from Node.JS these additional headers being added and being unavoidable for existing code / requests is IMO a breaking change.
> 
> We could make this not-breaking by not inferring the `origin` option via the `origin` header but even then I'd want to test all existing requests to ensure we haven't accidentally started adding random headers to existing functioning requests.
> 
> TLDR: My stance is that adding unexpected and unremovable headers to a request is a semver/major change

Thanks for your reply. I guess I didn't explain it clearly. The commit doesn't make the net.request sends additional headers.
This statement is not true:
```This PR adjusts the behavior of an existing option (the headers object) to send unexpected additional headers if you provide an origin header.```

Let me list what really happened in  a realworld app.

* case 1: no origin
```
net.request(url)
```
* case 2: the same origin is specified.
```
net.request(url, { origin: url })
```
* case 3: an origin other then url is specified.
```
net.request(url, { origin: url })
```

| Case      | Headers(without #e1cc78)  | Headers(with #e1cc78) |
| ----------- | ----------- | ----------- |
| 1      | <ul> <li>'Sec-Fetch-Dest': 'empty' </li> <li> 'Sec-Fetch-Mode': 'no-cors'</li>  <li>'Sec-Fetch-Site': 'none'</li> </ul>         | <ul> <li>'Sec-Fetch-Dest': 'empty' </li> <li> 'Sec-Fetch-Mode': 'no-cors'</li>  <li>'Sec-Fetch-Site': 'none'</li> </ul> |
| 2   | <ul> <li>'Sec-Fetch-Dest': 'empty' </li> <li> 'Sec-Fetch-Mode': 'no-cors'</li>  <li>'Sec-Fetch-Site': 'none'</li> </ul>  | <ul> <li>'Sec-Fetch-Dest': 'empty' </li> <li> 'Sec-Fetch-Mode': 'no-cors'</li>  <li>'Sec-Fetch-Site': 'same-site'</li> </ul> |
| 3   | <ul> <li>'Sec-Fetch-Dest': 'empty' </li> <li> 'Sec-Fetch-Mode': 'no-cors'</li>  <li>'Sec-Fetch-Site': 'none'</li> </ul> | <ul> <li>'Sec-Fetch-Dest': 'empty' </li> <li> 'Sec-Fetch-Mode': 'no-cors'</li>  <li>'Sec-Fetch-Site': 'cross-site'</li> </ul>  |

These headers are always sent by chromium networking stack. The commit just change the value of Sec-Fetch-* headers to match behivour of fetch() but **doesn't** really send any new headers. Without the commit the networking always send headers in the second column, and you couldn't change any of Sec-Fetch-* headers since they're hardcode in the c++ side.

Unexpected headers sent without the commit, this is the reason why I send the PR and what I intent to fix.

case 4: 
```
net.request(url, {
Origin: url,
'Sec-Fetch-Dest': 'script',
'Sec-Fetch-Mode': 'no-cors',
'Sec-Fetch-Site': 'same-origin'
})
```
Headers sent:
```
...
'Sec-Fetch-Dest': 'empty',
'Sec-Fetch-Mode': 'no-cors',
'Sec-Fetch-Site': 'none'
```

**Comment by LuoJinghua** [comment]
Or could we consider backport these changes and bump the version number?

**Comment by LuoJinghua** [comment]
Is there a reason why the `backport/requested` label is removed?

**Comment by nornagon** [comment]
This is now no longer relevant.
